xblock/scorable.py

```diff
@@ -0,0 +1,158 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import six
+
+
+Score = namedtuple('Score', ['earned', 'total'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Provides the following public methods:
+
+        ScorableXBlockMixin.rescore(self, only_if_higher)
+            Calculate a new score and save it to the block.
+            `only_if_higher
+
+        ScorableXBlockMixin.allows_rescore(self) -> bool
+            Returns True if the block can be rescored.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.get_score(self)
+            Return the currently saved score.  Must not calculate a new score.
```

Comment:
@sanfordstudent I remember reviewing the drag-n-drop code in a previous PR earlier where this came up.  Are you sure they store their score as a weighted value?  That sounds like a significant issue - since a problem weight change normally doesn't require a re-storing of all scores.